xds/src/main/java/io/grpc/xds/sds/SdsClient.java

```diff
@@ -0,0 +1,308 @@
+
+package io.grpc.xds.sds;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import io.envoyproxy.envoy.api.v2.DiscoveryResponse;
+import io.envoyproxy.envoy.api.v2.auth.SdsSecretConfig;
+import io.envoyproxy.envoy.api.v2.auth.Secret;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource;
+import io.envoyproxy.envoy.api.v2.core.ApiConfigSource.ApiType;
+import io.envoyproxy.envoy.api.v2.core.ConfigSource;
+import io.envoyproxy.envoy.api.v2.core.GrpcService;
+import io.envoyproxy.envoy.api.v2.core.GrpcService.GoogleGrpc;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.envoyproxy.envoy.service.discovery.v2.SecretDiscoveryServiceGrpc;
+import io.envoyproxy.envoy.service.discovery.v2.SecretDiscoveryServiceGrpc.SecretDiscoveryServiceStub;
+import io.grpc.Internal;
+import io.grpc.ManagedChannel;
+import io.grpc.Status;
+import io.grpc.netty.NettyChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.EpollDomainSocketChannel;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+import io.netty.channel.unix.DomainSocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * SDS client used by {@link SslContextProvider}s to get SDS updates from the SDS server. This is
+ * most likely a temporary implementation until merged with the XdsClient.
+ */
+// TODO(sanjaypujare): once XdsClientImpl is ready, merge with it and add retry logic
+@Internal
+final class SdsClient {
+  private static final Logger logger = Logger.getLogger(SdsClient.class.getName());
+  private static final String SECRET_TYPE_URL = "type.googleapis.com/envoy.api.v2.auth.Secret";
+
+  private final HashMap<String, HashSet<SecretWatcher>> watcherMap = new HashMap<>();
+
+  private static SdsClient instance;
+  private final ConfigSource configSource;
+  private final Node clientNode;
+  @VisibleForTesting String targetUri;
+  ManagedChannel channel;
+  private SecretDiscoveryServiceStub secretDiscoveryServiceStub;
+  private ResponseObserver responseObserver;
+  private StreamObserver<DiscoveryRequest> requestObserver;
+  private DiscoveryResponse lastResponse;
+
+  /**
+   * Starts resource discovery with SDS protocol. This method should be the first one to be called
+   * and should be called only once.
+   */
+  void start() {
+    if (channel == null) {
+      if (targetUri.startsWith("unix:")) {
+        EventLoopGroup elg = new EpollEventLoopGroup();
```

Comment:
I just discovered we will have to use more than one client :-( because the SDS server (Node Agent) does not support more than one resource per SDS stream. I think the best thing to do is to support multiple clients (one client per resource name). Will look into ELG/number of threads and epoll availabilty